test/rules/no_self_assign_test.dart

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:test_reflective_loader/test_reflective_loader.dart';
+
+import '../rule_test_support.dart';
+
+main() {
+  defineReflectiveSuite(() {
+    defineReflectiveTests(NoSelfAssignTest);
+  });
+}
+
+@reflectiveTest
+class NoSelfAssignTest extends LintRuleTest {
+  @override
+  String get lintRule => 'no_self_assign';
+
+  test_fieldAssignment() async {
+    await assertDiagnostics(r'''
+class C {
+  int x = 5;
+
+  C(int x) {
+    x = x;
+  }
+}
+''', [lint(41, 5)]);
+  }
+
+  test_fieldAssignmentDifferentVar() async {
+    await assertNoDiagnostics(r'''
+class C {
+  int x = 5;
+
+  C(int y) {
+    x = y;
+  }
+}
+''');
+  }
+
+  test_fieldAssignmentExplicit() async {
+    await assertNoDiagnostics(r'''
+class C {
+  int x = 5;
+
+  C(int x) {
+    this.x = x;
+  }
+}
+''');
+  }
+
+  test_fieldAssignmentExplicitSameVar() async {
+    await assertDiagnostics(r'''
+class C {
+  int x = 5;
+
+  void update(C other) {
+    other.x = other.x;
+  }
+}
+''', [lint(53, 17)]);
+  }
+
+  test_fieldInitialization() async {
+    await assertNoDiagnostics(r'''
+class C {
+  int x;
+
+  C(int x) : x = x;
+}
+''');
+  }
+
+  test_propertyAssignment() async {
+    await assertDiagnostics(r'''
+class C {
+  int _x = 5;
+
+  int get x => _x;
+
+  set x(int x) {
+    _x = x;
+  }
+
+  void example() {
+    x = x;
+  }
+}
+''', [lint(102, 5)]);
+  }
+
+  test_classMemberAssignment() async {
+    await assertDiagnostics(r'''
+class C {
+  static String foo = "foo";
+}
+
+void main() {
+  C.foo = C.foo;
+  String foo;
+  foo = C.foo; // OK
```

Comment:
Would you mind splitting out the lint case and the OK case into two tests? It hurts me to count 58 characters to see what it is reported :)

Reply:
Done